app/support/auto_expire.rb

```diff
@@ -0,0 +1,44 @@
+class AutoExpire
+  def perform
+    order_details.each do |od|
+      od.transaction do
+        expire_reservation(od)
+      end
+    end
+  end
+
+  private
+
+  def order_details
+    purchased_active_order_details | non_reservation_order_details
+  end
+
+  def purchased_active_order_details
+    OrderDetail.purchased_active_reservations
+      .where("reservations.reserve_end_at < ?", Time.zone.now - 12.hours)
+      .readonly(false)
+      .all
+  end
+
+  def non_reservation_order_details
+    OrderDetail.purchased_active_reservations
+      .where("reservations.reserve_end_at < ?", Time.zone.now)
+      .joins(:product)
+      .merge(Instrument.reservation_only)
+      .readonly(false)
+  end
+
+  def expire_reservation(od)
+    od.complete!
+    od.fulfilled_at = od.reservation.reserve_end_at
```

Comment:
I think this should probably happen before the `complete!` call so it gets saved as part of that call. Otherwise, the fulfilled_at might not get set if the price policy is missing.